coreutils: Fix build on native FreeBSD

rhelmot opened the pull request:
## Description of changes

part of https://github.com/NixOS/nixpkgs/pull/296581

The singleBinary diff is so old and has been tossed through so many commits that it I no longer remember the nature of the horrors that are unleashed without it, but it is required last I checked. The configure diff is explained inline. The comment moving is a formatting silly and can be reverted if it is not appropriate.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] x86_64-freebsd
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

alyssais commented on pkgs/tools/misc/coreutils/default.nix:
```diff
@@ -16,7 +16,7 @@
 , withOpenssl ? !minimal, openssl
 , withPrefix ? false
-, singleBinary ? "symlinks" # you can also pass "shebangs" or false
+, singleBinary ? if stdenv.isFreeBSD then false else "symlinks" # you can also pass "shebangs" or false
```
Does shebangs work?

rhelmot replied:
looks like it does, thanks for the idea!

rhelmot replied:
Looks like it does not work. FreeBSD doesn't seem to support using interpreted scripts as interpreters, so anything with coreutils/bin/env as the interpreter will fail. I'm debugging symlinks right now to see what the deal is.

rhelmot replied:
Best I can tell, coreutils is busted for the single-binary build, and freebsd just happens to trigger it. [This](https://github.com/coreutils/coreutils/blob/master/src/local.mk#L485) seems to be the root cause - mixing DEPENDENCIES (i.e. make target prerequisites) and LDADD (i.e. flags for ld) is a really bad idea, since suddenly you have -lintl as a make target.

I don't know why shebangs was working two days ago. Maybe suddenly stripping all our libs causes coreutils configure to detect different LDADD flags?

I'm going to revert to singleBinary = false, submit an upstream issue, link to said upstream issue, and move on with my life.

rhelmot commented on pkgs/tools/misc/coreutils/default.nix:
```diff
@@ -138,7 +138,10 @@ stdenv.mkDerivation rec {
+    ++ optional stdenv.hostPlatform.isLinux "gl_cv_have_proc_uptime=yes"
+    # this function exists in libc but is not provided by any header, leading
+    # to its detection at configure flag and failure at build time, I think
+    ++ optional stdenv.hostPlatform.isFreeBSD "ac_cv_func_random_r=no";
```
No, on non-Nix builds this value is detected automatically. I'm really not sure what could cause this

alyssais replied:
Might be worth comparing config.log between Nix and non-Nix builds?

rhelmot replied:
Without a doubt, chasing this rabbit hole was the most awful experience I've had debugging freebsd/nix so far. Our libc had a bunch of local symbols in it, which are usually removed by stripping. These are what was being picked up by coreutils configure to believe that random_r existed. The reason they weren't being stripped was that *our install wrapper purposefully discarded the flag that would cause it to invoke strip*. The nixpkgs blame goes back to the original init freebsd commit. I can't for the life of me figure out why this would be considering that the same commit specifies that that flag should be passed! I suppose it has fallen to me to debug this weighty error. Everyone say, "Thank you, Atlas". Fix pushed.

alyssais replied:
Thank you so much for the wonderful investigation!

alyssais reviewed (commented):
Can you rebase to have one commit fixing the strip wrapper, then one commit fixing coreutils?  Then LGTM.

alyssais commented on pkgs/tools/misc/coreutils/default.nix:
```diff
@@ -98,8 +98,8 @@ stdenv.mkDerivation rec {
+  # autoreconfHook is due to patch, normally only needed for cygwin
   nativeBuildInputs = [
-    # autoreconfHook is due to patch, normally only needed for cygwin
     autoreconfHook
```
This hunk can be dropped.

rhelmot committed:
freebsd: Don't discard request to strip installed binaries

rhelmot commented on pkgs/tools/misc/coreutils/default.nix:
```diff
@@ -99,13 +99,12 @@ stdenv.mkDerivation rec {
   ]
   ++ optionals stdenv.hostPlatform.isCygwin [
     # due to patch
+    autoreconfHook
```
as per https://github.com/NixOS/nixpkgs/pull/311789/files/977ab5018bd37332386fd36af9227b7fa7d42eab#r1601167922

alyssais replied:
Separate commit?  Doesn't really belong in one called "disable single-binary on native FreeBSD".

alyssais commented on pkgs/tools/misc/coreutils/default.nix:
```diff
@@ -16,7 +16,7 @@
 , withOpenssl ? !minimal, openssl
 , withPrefix ? false
-, singleBinary ? "symlinks" # you can also pass "shebangs" or false
+, singleBinary ? if stdenv.isFreeBSD then false else "symlinks" # you can also pass "shebangs" or false
```
Link to upstream issue, either here or in commit message?

rhelmot replied:
Is there a coreutils bug tracker other than the mailing list? I emailed them like 12h ago and it still hasn't shown up in their archives for a link.

alyssais replied:
To bug-coreutils@gnu.org?  Did you send it in plain text?

rhelmot replied:
I did

alyssais replied:
I guess it probably got held for moderation/spam review then?  You could mention the message ID and date of the message, and then if it ever does show up it'll be findable.

rhelmot replied:
Looks like it got approved overnight. It's in the commit message now.

rhelmot committed:
coreutils: disable single-binary on native FreeBSD

Upstream bug report:
https://lists.gnu.org/archive/html/bug-coreutils/2024-05/msg00033.html

rhelmot committed:
coreutils: make autoreconfHook dependency conditional
